api_core/google/api_core/iam.py

```diff
@@ -79,39 +121,120 @@ class Policy(collections_abc.MutableMapping):
     def __init__(self, etag=None, version=None):
         self.etag = etag
         self.version = version
-        self._bindings = collections.defaultdict(set)
+        self._bindings = []
 
     def __iter__(self):
-        return iter(self._bindings)
+        self.__check_version__()
+        return (binding["role"] for binding in self._bindings)
 
     def __len__(self):
+        self.__check_version__()
         return len(self._bindings)
 
     def __getitem__(self, key):
-        return self._bindings[key]
+        self.__check_version__()
+        for b in self._bindings:
+            if b["role"] == key:
+                return b["members"]
+        return set()
 
     def __setitem__(self, key, value):
-        self._bindings[key] = set(value)
+        self.__check_version__()
+        value = set(value)
+        for binding in self._bindings:
+            if binding["role"] == key:
+                binding["members"] = value
+                return
+        self._bindings.append({"role": key, "members": value})
 
     def __delitem__(self, key):
-        del self._bindings[key]
+        self.__check_version__()
+        for b in self._bindings:
+            if b["role"] == key:
+                self._bindings.remove(b)
+                return
+        raise KeyError(key)
+
+    def __check_version__(self):
+        """Raise InvalidOperationException if version is greater than 1 or policy contains conditions."""
+        raise_version = self.version is not None and self.version > 1
+
+        if raise_version or self._contains_conditions():
+            raise InvalidOperationException(_DICT_ACCESS_MSG)
+
+    def _contains_conditions(self):
+        for b in self._bindings:
+            if b.get("condition") is not None:
+                return True
+        return False
+
+    @property
+    def bindings(self):
+        """:obj:`list` of :obj:`dict`: The policy's bindings list.
```

Comment:
Not sure how we'd document `dict` keys, I couldn't find any examples in the Google style guides.

Reply:
[Mypy](http://mypy-lang.org/examples.html) uses the form `Dict[key_type, value_type]`. Not sure if this is an official guideline, but at least in `BigQuery` we follow this format. I also found quite a few occurrences  of that in `firestore`.

Hope this helps!